src/styles/main.scss

```diff
@@ -0,0 +1,97 @@
+@import './utils/variables';
+
+$padding: 10px;
+$gap: 1px;
+$total-gap: $gap * 8;
+$square-side: 100px;
+$line-width: ($square-side * 7) + ($padding * 2) + $total-gap;
+$day-of-the-week: (
+  'mon': 0,
+  'tue': (
+    ($square-side + $gap),
+  ),
+  'wed': (
+    ($square-side + $gap) * 2,
+  ),
+  'thu': (
+    ($square-side + $gap) * 3,
+  ),
+  'fri': (
+    ($square-side + $gap) * 4,
+  ),
+  'sat': (
+    ($square-side + $gap) * 5,
+  ),
+  'sun': (
+    ($square-side + $gap) * 6,
+  ),
+);
+
+body {
+  margin: 0;
+}
+
+.wrapper {
+  width: 100vw;
+  height: 100vh;
+  display: flex;
+  justify-content: center;
+  align-items: center;
+}
+
+.calendar {
+  display: flex;
+  flex-wrap: wrap;
+  gap: $gap;
+  box-sizing: border-box;
+  padding: $padding;
+  width: $line-width;
+
+  transform: translateX(1px);
+
+  &__day {
+    position: relative;
+    display: none;
+    box-sizing: border-box;
+    padding: 0;
+    margin: 0;
+    width: $square-side;
+    height: $square-side;
+    background-color: $grey;
+    border: 1px solid #000;
+    transition: transform 0.5s;
+    &:hover {
+      cursor: pointer;
+      background-color: $pink;
+      transform: translateY(-20px);
+    }
+  }
+}
+
+@for $i from 28 through 31 {
```

Comment:
your approach is not bad but there is better:

variables:


```
$calendar-cell-size: 100px;
$calendar-gap-size: 1px;
$calendar-padding: 10px;
$calendar-days: (
  mon: 0,
  tue: 1,
  wed: 2,
  thu: 3,
  fri: 4,
  sat: 5,
  sun: 6,
);
```
code:

```
  &__day {
    @for $i from 1 through 31 {
      &:nth-child(#{$i})::before {
        content: '#{$i}';
      }
    }
  }
```
```

  @each $calendar-day, $calendar-index in $calendar-days {
    &--start-day-#{$calendar-day} :first-child {
      margin-left: calc(
        #{$calendar-index} * $calendar-cell-size + #{$calendar-index} * $calendar-gap-size
      );
    }
  }

  @for $length from 28 through 31 {
    &--month-length-#{$length} :nth-child(n + #{$length + 1}) {
      display: none;
    }
  }

```

Reply:
If you don't mind I will leave my variant) 
Your approach is easier to read) 

But it was a challenge for me to use loop inside loop correctly here) 
It is hard to refuse my attempt)